garage/tf/policies/maml_policy.py

```diff
@@ -0,0 +1,151 @@
+"""Maml Policy."""
+import tensorflow as tf
+
+from garage.tf.policies.base2 import StochasticPolicy2
+
+# This is an alternative variable store for maml.
+MAML_VARIABLE_STORE = set()
+MAML_TASK_INDEX = 0
+
+
+class MamlPolicy(StochasticPolicy2):
+    """
+    MamlPolicy used by Model-Agnostic Meta-Learning.
+
+    This policy take any garage policies that use models and augment
+    them with one step adaptation tensors.
+
+    Args:
+        wrapped_policy: A garage policy using model.
+        n_tasks: Number of task for meta learning.
+        adaptation_step_size: The step size for one step adaptation.
+        name: Name of this policy.
+    """
+
+    def __init__(self,
+                 wrapped_policy,
+                 n_tasks,
+                 adaptation_step_size=0.5,
+                 name="MamlPolicy"):
+
+        self.wrapped_policy = wrapped_policy
+        self.n_tasks = n_tasks
+        self.name = name
+        self._initialized = False
+        self._adaptation_step_size = adaptation_step_size
+        self._adapted_param_store = dict()
+        self._create_update_opts()
+
+        super().__init__(wrapped_policy._env_spec)
+
+    def initialize(self, gradient_var, inputs):
+        """
+        Initialize the MAML Policy.
+
+        This funtion will create all the one step adapted parameters
+        and rebuild a model using these adapted parameters for each
+        task. A MamlPolicy can only be initialized once.
+
+        Args:
+            gradient_var: Gradient variables for one step adaptation.
+            inputs: Input tensors for rebuilding the models.
+
+        Returns:
+            all_model_outputs: The outputs of rebuilt models.
+            update_opts: Adapted parameter tensors.
+            update_opts_input: Input for calculating adapted parameters.
+        """
+
+        assert not self._initialized, "The MAML policy is initialized and can be initialized once."
+
+        global MAML_VARIABLE_STORE, MAML_TASK_INDEX
+        update_opts = []
+        with tf.name_scope(self.name):
+            # One step adaptation
+            with tf.name_scope("Adaptation"):
+                for i in range(self.n_tasks):
+                    params = self.wrapped_policy.get_params()
+                    gradient_i = gradient_var[i]
+
+                    for p, g in zip(params, gradient_i):
+                        adapted_param = p - self._adaptation_step_size * g
+                        name = "maml_policy/{}/{}".format(i, p.name)
+                        self._adapted_param_store[name] = adapted_param
+                        if i == 0:
+                            update_opts.append(adapted_param)
+
+            def maml_get_variable(name, shape=None, **kwargs):
+                scope = tf.get_variable_scope()
+                idx = 0
+                if MAML_TASK_INDEX >= self.n_tasks:
+                    raise ValueError("Invalid TASK_INDEX.")
+                fullname = "maml_policy/{}/{}/{}:{}".format(
```

Comment:
perhaps you can make some helper(s) like `_get_name` to replace all these arbitrary string interpolations.

you're actually trying to create an entire system of unique paths. it would be worth it to create a simple helper class/functions which you can test to make sure they satisfy your expectations for uniqueness and other properties.